src/snapshot/component_checksum_hash.rs

```diff
@@ -0,0 +1,67 @@
+use std::{
+    hash::{Hash, Hasher, BuildHasher},
+    marker::PhantomData,
+};
+
+use bevy::prelude::*;
+
+use crate::{ChecksumFlag, ChecksumPart, Rollback, SaveWorld, SaveWorldSet};
+
+/// A [`Plugin`] which will track the [`Component`] `C` on [`Rollback Entities`](`Rollback`) and ensure a
+/// [`ChecksumPart`] is available and updated. This can be used to generate a [`Checksum`](`crate::Checksum`).
+pub struct GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    _phantom: PhantomData<C>,
+}
+
+impl<C> Default for GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    fn default() -> Self {
+        Self {
+            _phantom: default(),
+        }
+    }
+}
+
+impl<C> GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    /// A [`System`] responsible for managing a [`ChecksumPart`] for the [`Component`] type `C`.
+    pub fn update(
+        mut commands: Commands,
+        components: Query<(&Rollback, &C), (With<Rollback>, Without<ChecksumFlag<C>>)>,
+        mut checksum: Query<&mut ChecksumPart, (Without<Rollback>, With<ChecksumFlag<C>>)>,
+    ) {
+        let mut hasher = bevy::utils::FixedState::default().build_hasher();
+
+        let mut components = components.iter().collect::<Vec<_>>();
+
+        components.sort_by_key(|(&rollback, _)| rollback);
```

Comment:
Ok I spent some time and I think I have a good solution to this problem. What I've done is added a resource, `RollbackOrdered`, which contains a sorted list of every `Rollback` component ever created. This list is sorted and updating during `AddRollbackCommand` execution, which should avoid any runtime penalty, since all rollback entities should be registered during times when it is reasonable to be spawning entities anyway.

Doing it here allows me to add the index of the `Rollback` flag to the hash for each value in a component hash, making it unique for that component-rollback pairing. Then I can combine all those unique values using XOR, which is commutative (order independent).

The result is a checksum which is order stable, resistant to transposition, and doesn't require sorting.